### PR TITLE
Create C# implementation of Exponentiation using Squaring

### LINKED ISSUE #602: Exponentiation by Squaring in C#
This is an open issue for the Exponentiation by Squaring algorithm implementation in C#. If you wish to contribute, just make some noise and we will assign you.

### PR DESCRIPTION
**Pull request by MrLukeKR**
**Fixes #602**

By submitting this pull request I confirm I've read and complied with the below declarations.

-   [x] I have read the [Contribution guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md) 
and I am confident that my PR reflects them.

-   [x] I have followed the [coding guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#cs) 
for this project.

-   [x] My code follows the [skeleton code structure](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#sample).

-   [x] This pull request has a descriptive title. 
For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.

-   [x] This pull request shall only be merged once the build passes.

-   [x] I have mentioned the issue number correctly (with hyper-link) in this pull request description.

### CODE REVIEW

File: algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+using System;
+
+public class ExponentiationBySquaring
```
Comment (RavinduSachintha) [inline algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs:3]: Codacy found an issue: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6352723)

File: algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+{
+	public static double exponentiationBySquaring(double base_num, double exponent){
+		if(exponent == 0)
+		{
+			// If the exponent is 0, the answer is always 1
+			return 1;
+		}
+		else if(exponent % 2 == 0)
```
Comment (RavinduSachintha) [inline algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs:11]: Codacy found an issue: [Do not check floating point equality with exact values, use a range instead.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6352723)

File: algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+
+	/*
+		In the main function, run an example of the exponentiation algorithm
+		using 5 ^ 8 (5 to the power of 8)
+	*/
+	public static void Main()
+	{
+		double val = exponentiationBySquaring(5, 8);
+		Console.WriteLine(val);
```
Comment (RavinduSachintha) [inline algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs:40]: Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6352723)

File: algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+using System;
+
+public class ExponentiationBySquaring
+{
+	public static double exponentiationBySquaring(double base_num, double exponent){
+		if(exponent == 0)
```
Comment (RavinduSachintha) [inline algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs:6]: Codacy found an issue: [Do not check floating point equality with exact values, use a range instead.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6352723)

File: algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	}
+
+
+	/*
+		In the main function, run an example of the exponentiation algorithm
+		using 5 ^ 8 (5 to the power of 8)
+	*/
+	public static void Main()
+	{
+		double val = exponentiationBySquaring(5, 8);
```
Comment (RavinduSachintha) [inline algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs:39]: Codacy found an issue: [Remove the unused local variable 'val'.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6352723)

File: algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+
+	/*
+		In the main function, run an example of the exponentiation algorithm
+		using 5 ^ 8 (5 to the power of 8)
+	*/
+	public static void Main()
+	{
+            var value = exponentiationBySquaring(5, 8);
+            Console.WriteLine("5^8 = " + value);
```
Comment (RavinduSachintha) [inline algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs:40]: Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6352723)

Comment (MrLukeKR) [reply algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs:40]: The code skeletons in the contributing guide show use of Console.WriteLine to demonstrate the algorithm; shall I keep the code the same as it is here, or make the main function blank?

Comment (RavinduSachintha) [inline algorithms/ar-expsq/csharp/ExponentiationBySquaring.cs:40]: Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6352723)

**Comment by MrLukeKR** [comment]
When merging, could you label with hacktoberfest-accepted? Many thanks :)

**Comment by RavinduSachintha** [comment]
> When merging, could you label with hacktoberfest-accepted? Many thanks :)

sure. we will do. before that, would you be able to add this contribution to the algorithm cart in the readme file? don't change anything else.

**Comment by MrLukeKR** [comment]
Sure thing :)

**Review by RavinduSachintha — APPROVED** [review]

**Comment by RavinduSachintha** [comment]
@all-contributors please add @MrLukeKR for code and doc

**Comment by RavinduSachintha** [comment]
@MrLukeKR thank you for your contribution. keep it up :+1: 
make sure to star this repo :wink:

**Comment by MrLukeKR** [comment]
Will do, thanks very much 😁 can I do the python2 Dijkstra now? 😂😂😉

**Pull request merged**